[RFC] Bootstrap ta

jenswi-linaro opened the pull request:
This pull request introduces bootstrap TAs which can completely replace the current TAs depending on how we'd like to progress.

I see some ways forward:
1. Add this as yet another method of loading TAs
2. Replace the current way of loading TAs from supplicant with this
3. Wait until we have some security domain in place before starting to retire/deprecate 1

Doing 2 (or 3) doesn't mean that we have to remove the old code, we could just make it optional. But we'd change our current test setup to only install TAs via bootstrap TAs (or via a security domain).

Bootstrap TA are basically only the solution to the chicken and egg problem for security domains. Next step would be to introduce a security domain.

jenswi-linaro commented:
Uses https://github.com/OP-TEE/optee_test/pull/240 to install bootstrap TAs

Note that the 9000 tests will fail if bootstrap TAs are installed.

lorc commented:
Hi @jenswi-linaro, just a quick question: what do you mean when you say "security domain"?

lorc commented on scripts/sign_bsta.py:
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python
```
Have you tested this script in both python2 and python3?

jenswi-linaro replied:
I have now, and it seems to work. :-)

lorc replied:
Okay then :-)

lorc commented on scripts/sign_bsta.py:
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python
+#
+# Copyright (c) 2015, Linaro Limited
```
2017, I think :)

lorc commented on scripts/sign_bsta.py:
```diff
@@ -0,0 +1,77 @@
+#
+
+def uuid_parse(s):
+	from uuid import UUID
```
Please check this script with `pycodestyle` (or with `pip8.py`). There will be a lot of warnings.

jenswi-linaro commented:
Hi @lorc, this for instance: https://tools.ietf.org/html/draft-pei-opentrustprotocol-04#section-5.3

Note that Security Domains doesn't to have provide isolation between TAs in different Security Domains. A Security Domain is just a TA management domain.

lorc commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,702 @@
+
+	if (db->files)
+		return TEE_SUCCESS;
+
```
Remove extra empty line.

lorc commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,702 @@
+
+	assert(db->refc);
+	db->refc--;
+	if (db->refc)
```
How you can guarantee thread safety in tadb manipulation?

jenswi-linaro replied:
These functions are supposed to be from a static TA which will guarantee thread safety.

jenswi-linaro replied:
I just realized that these functions are used from `core/arch/arm/kernel/tadb_store.c` which is outside the pseudo TA context. Ideally I should have a read/write lock to allow concurrent loading of TAs.

lorc commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,702 @@
+static TEE_Result populate_files(struct tee_tadb_dir *db)
+{
+	TEE_Result res;
+	size_t idx;
```
Is it appropriate to use `size_t` for iterator variables? Maybe plain `int` (or `unsigned int`) will suit better?

jenswi-linaro replied:
I don't see any problem with it. `size_t` is usually an `unsigned long` which in its turn usually fits into a register so there's no overhead at least. The range of `size_t` is good for counting almost anything.

On the other hand, maybe I'm missing the point...

lorc commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,702 @@
+		res = set_file(db, entry.file_number);
+		if (res)
+			break;
+
```
Remove empty line.

lorc commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,702 @@
+	tadb_put(db);
+}
+
+static TEE_Result populate_files(struct tee_tadb_dir *db)
```
Purpose of this function is bit unclear to me...

jenswi-linaro replied:
I guess I thought this function was going to be used by several functions when doing the initial design. Now it's clear it's only needed in `tee_tadb_ta_create()`. I'll move the implementation into `tee_tadb_ta_create()` instead.

lorc commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,702 @@
+	size_t idx;
+	const TEE_UUID null_uuid = { 0 };
+
+	if (db->files)
```
There you return if `db->files != NULL`...

lorc commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,702 @@
+		if (!memcmp(&entry.prop.uuid, &null_uuid, sizeof(null_uuid)))
+			continue;
+
+		if (test_file(db, entry.file_number)) {
```
... and there you use `db->files` to check if `file_number` is stored. But at L265 you ensured that `db->files` is `NULL`.

jenswi-linaro replied:
As this is called in a loop, `db->files` could have been allocated at this point. `test_file()` and `set_file()` deals with that.

lorc commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,702 @@
+	int i = 0;
+
+	if (!ta)
+		res = TEE_ERROR_OUT_OF_MEMORY;
```
Looks like you missed `return` there.

jenswi-linaro replied:
Thanks

lorc commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,702 @@
+		size_t num_bytes = 0;
+
+		while (num_bytes < l) {
+			uint8_t b[1024 * 4];
```
Are you really sure, you want  to allocate 4kb on stack?

jenswi-linaro replied:
You're right, it's a bit excessive. I'll replace it with `malloc()`.

lorc commented on core/arch/arm/pta/management.c:
```diff
@@ -0,0 +1,288 @@
+	res = tee_hash_get_digest_size(TEE_DIGEST_HASH_TO_ALGO(shdr->algo),
+				       &hash_size);
+	if (res != TEE_SUCCESS)
+		return res;
```
`return TEE_ERROR_SECURITY`?

lorc commented on core/arch/arm/pta/management.c:
```diff
@@ -0,0 +1,288 @@
+
+	res = crypto_ops.acipher.alloc_rsa_public_key(&key, shdr->sig_size);
+	if (res != TEE_SUCCESS)
+		return res;
```
`return TEE_ERROR_SECURITY`?

lorc commented on core/arch/arm/pta/management.c:
```diff
@@ -0,0 +1,288 @@
+				     sizeof(*shdr));
+	if (res)
+		goto err_free_hash_ctx;
+
```
Remove extra empty line.

jenswi-linaro commented:
Addressed comments, except then one with mutex (waiting for the outcome of https://github.com/OP-TEE/optee_os/pull/1942)

lorc commented on core/tee/tadb.c:
```diff
@@ -367,6 +323,66 @@ static bool __maybe_unused tadb_check_state(const struct tee_tadb_ta *ta,
+			continue;
+
+		if (test_file(db, entry.file_number)) {
+			DMSG("Clearing duplicate file number %" PRIu32,
```
If that is not going to happen, then `EMSG` or at least `IMSG` will fit much better.

lorc commented on core/tee/tadb.c:
```diff
@@ -367,6 +323,66 @@ static bool __maybe_unused tadb_check_state(const struct tee_tadb_ta *ta,
+
+		res = read_ent(db, idx, &entry);
+		if (res) {
+			if (res == TEE_ERROR_ITEM_NOT_FOUND)
```
It is confusing, that only one successful (apart from L336) return from this function resides in the middle of it. 
Now, when I grasped idea of this function, this seems obvious. But it was hard to understand from a first try.

This is just a note. I don't ask you to perform any actions.

lorc commented on core/tee/tadb.c:
```diff
@@ -670,17 +694,25 @@ TEE_Result tee_tadb_ta_read(struct tee_tadb_ta *ta, void *buf, size_t *len)
 			return res;
 	} else {
 		size_t num_bytes = 0;
+		size_t b_size = MIN(SIZE_4K, l - num_bytes);
```
You zero out `num_bytes` at the line above. Better to leave only `MIN(SIZE_4k, l)`. It is easier to read.

jenswi-linaro commented:
Rebased and squashed on top of master to be able to use the read lock. It turned out atomic reference counting routines where needed too, added that to the beginning of the PR.

jenswi-linaro commented on core/include/kernel/refcount.h:
```diff
@@ -0,0 +1,27 @@
+	unsigned int val;
+};
+
+/* Increases refcount by 1, return true if val != 0 else false */
```
You're right. I'll update the description.

jenswi-linaro commented on core/kernel/refcount.c:
```diff
@@ -0,0 +1,38 @@
+	unsigned int oval = atomic_load_uint(&r->val);
+
+	while (true) {
+		nval = oval + 1;
```
That's actually taken care of by `atomic_cas_uint()`

etienne-lms replied:
ok, right, sorry.

etienne-lms commented on core/include/tee/tadb.h:
```diff
@@ -0,0 +1,72 @@
+ *		TA binary
+ * @bin_size:	Size of the binary TA
+ * @is_ta:	True if this is a TA
+ * @is_root:	True if this is a root SD or a TA without an SD
```
s/an/a/

jenswi-linaro replied:
It depends on how you pronounce it, I pronounce it ess-dee

etienne-lms replied:
ok :)

etienne-lms commented on core/include/tee/tadb.h:
```diff
@@ -0,0 +1,72 @@
+ */
+struct tee_tadb_property {
+	TEE_UUID uuid;
+	TEE_UUID parent;/* parent security domain, zeroes if nonexistent */
```
minor: tab before comment.

jenswi-linaro replied:
It doesn't fit on the line then. I'll remove it now that there's a description of the struct above.

etienne-lms commented:
minor: could you rename the pTA "management" into sometihng like "TA management"? For example **management.c** being a bit too generic, even in **pta/** subdir. Something like **ta_management.c**?

jenswi-linaro commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,762 @@
+#include <tee/tee_svc_storage.h>
+#include <utee_defines.h>
+
+#define TADB_MAX_BUFFER_SIZE	(64U * 1024)
```
We do need some value, 64KiB is nothing for normal world to allocate as shared memory.

jenswi-linaro commented:
Update

jforissier commented:
In order to give an opinion on this RFC, I need to understand the big picture. How does it work (high-level)? What problem are you trying to address? What is different from what we are currently doing? What is a bootstrap TA? Why does it have a different extension as mentioned in the `optee_test` PR (`*.bsta`)? How do you generate a `.bsta`? Is a different signing key required? Are bootstrap TAs still stored in the REE FS? You're mentioning security domains and how this could be a solution to the chicken-and-egg problem SDs have. How?

Sorry I don't feel like trying to figure this out from code only :(

jenswi-linaro commented:
Some description available here: 
https://docs.google.com/document/d/1fIZkKNjSkaVjN0QsuJBxtJsU3T5GjtSayC-eXklm47Q/edit?usp=sharing

jforissier commented on core/include/kernel/refcount.h:
```diff
@@ -0,0 +1,71 @@
+
+/* Increases refcount by 1, return true if val > 0 else false */
+bool refcount_inc(struct refcount *r);
+/* Decreases refcount by 1, return true if val > 0 else false */
```
`return true if val == 0`

jforissier commented on core/include/kernel/refcount.h:
```diff
@@ -0,0 +1,71 @@
+ * Reference conter
+ *
+ * When val becomes 0 refcount_inc() refuses to change the value any longer
+ * and returns false.
```
I find the wording a bit weird, a more neutral tone may be better: "When val is 0, `refcount_inc()` does not change the value and returns false.". You may add "Otherwise, it increments the value and returns true".

jforissier commented on core/tee/sub.mk:
```diff
@@ -40,6 +40,9 @@ srcs-y += tee_obj.c
 srcs-y += tee_pobj.c
 srcs-y += tee_time_generic.c
 
+$(call force,CFG_SECURE_STORAGE_TA,y,required by CFG_MANAGEMENT_PTA)
```
Not needed here

jenswi-linaro commented:
Squashed and inserted some extra commit but not rebased to make it easier to compare with the old a2aa4a022849ec11ba4a6d2bbfa2b3bbdb3f53fa state.

History is reordered to instead of adding a new sign script, the old one is replaced/updated and there's no .bsta file any longer. The REE FS TA storage now supports the new TA format as well as the old format.

I missed addressing the refcount comments, will make an update commit later. Documentation will also follow later.

jenswi-linaro commented:
Fixes problem with old filenumbers not being reused.

jenswi-linaro commented:
Addresses the refcount comments

jenswi-linaro commented:
Added documentation in #1986

jforissier commented on scripts/sign.py:
```diff
@@ -1,80 +1,81 @@
+    from argparse import ArgumentParser
+
+    parser = ArgumentParser()
+    parser.add_argument('--uuid', required=True,
```
Could be optional (default: parse --in file name).

jenswi-linaro replied:
On the other hand, the name of the TA file doesn't have to be an UUID. Sounds like complicated usage to me.

jforissier replied:
True

jforissier commented on ta/arch/arm/link.mk:
```diff
@@ -58,4 +58,5 @@ $(link-out-dir)/$(binary).stripped.elf: $(link-out-dir)/$(binary).elf
 				$(TA_SIGN_KEY)
 	@echo '  SIGN    $@'
-	$(q)$(SIGN) --key $(TA_SIGN_KEY) --in $< --out $@
+	$(q)$(SIGN) --key $(TA_SIGN_KEY) --uuid $(binary) --version 0 \
```
Assuming `--uuid` is optional, the command could remain unchanged

jenswi-linaro commented on core/tee/tadb.c:
```diff
@@ -0,0 +1,760 @@
+	size_t rl = len;
+	struct tee_fs_rpc_operation op;
+
+	assert(tadb_check_state(ta, TEE_MODE_ENCRYPT));
```
This is an API usage error (programmers error). In the case of `tee_tadb_ta_write()`, `tee_tadb_ta_read()` and `tee_tadb_ta_close_and_commit()` it's possible to return an error, but the two functions `tee_tadb_ta_close_and_delete()`, `tee_tadb_ta_close()` are void functions.

In all of the cases there's no good cleanup action that the caller can perform.There's obviously some confusion regarding what the handle is to be used for. Initially I was using two different structs for the "read" class and "write" class of functions. Perhaps I should reintroduce that instead?

jforissier replied:
If there's no cleanup possible, then indeed a return code is useless.

> Initially I was using two different structs for the "read" class and "write" class of functions. Perhaps I should reintroduce that instead?

Yes, I think I would like it better.

jforissier commented:
How about s/bootstrap/secure storage/g? My reasoning is: bootstrapping another TA management method (OTrP, etc.) is only one way this "secure storage TA" feature may be used. It is also perfectly valid to install all TAs in secure storage. So it would be more neutral and more general to use "secure storage TA", `shdr_secstor_ta`, and so on.

jforissier commented on core/include/tee/tadb.h:
```diff
@@ -0,0 +1,72 @@
+	uint32_t custom_size;
+	uint32_t bin_size;
+	bool is_ta;
+	bool is_root;
```
Since `parent`, `handler`, `is_ta` and `is_root` are currently not used, I'd rather introduce them later. It is fine to mention in the commit log that this may be a first step towards supporting SDs, but it should not show in the code yet. The "secure storage TA" feature is useful on its own.

jenswi-linaro replied:
OK

jforissier commented:
@jenswi-linaro 

> I don't agree with s/bootstrap/secure storage/g.
> [...]

OK, fair enough. Personally I do think a number of users may just switch to the bootstrap TA format just for rollback protection. But, I agree it should not stop us from using the term "bootstrap".

The explanations/examples you're giving here and in the Google doc are quite valuable, and I think they should end up in the documentation folder or in a commit log (well, probably in both places, the commit log being less verbose obviously).

jenswi-linaro commented:
If they want to install a TA via PTA_SECSTOR_TA_MGMT they have to use the bootstrap format. All other use of the format is just to avoid using two formats in parallel.

There's obviously no roolback protection unless the TA actually is installed in secure storage. TAs signed in with the bootstrap format, but loaded via REE FS storage will not have any roolback protection.

Once the TA is installed in secure storage the bootstrap format is also not used any longer, so there really is only a small window where it's actually used.

jenswi-linaro commented:
The examples here and in the doc is mostly the direction I'm aiming for, but that may change as we learn more. I think we should add documentation as new features are added, not before.

I think I'll use the document as a living document more or less describing the direction, I'll add some comments from this PR to keep it in one place.

jenswi-linaro commented:
Update

jenswi-linaro commented:
Update

jforissier commented on mk/config.mk:
```diff
@@ -262,6 +262,14 @@ CFG_GP_SOCKETS ?= y
+# Enable storage for TAs in secure storage, depends on CFG_REE_FS=y
+# TA binaries are stored encrypted in the REE FS and are protected by
+# metadata in secure storage.
+CFG_SECSTOR_TA ?= y
```
Build fails at link time when `make CFG_REE_FS=n CFG_RPMB_FS=y` (this combination is tested by Shippable). I suggest the following to deal with the dependencies:

```diff
diff --git a/core/core.mk b/core/core.mk
index b633425022d2..e2640b3e0b46 100644
--- a/core/core.mk
+++ b/core/core.mk
@@ -20,12 +20,6 @@ PLATFORM_FLAVOR_$(PLATFORM_FLAVOR) := y
 
 $(call cfg-depends-all,CFG_PAGED_USER_TA,CFG_WITH_PAGER CFG_WITH_USER_TA)
 
-$(call cfg-depends-all,CFG_SECSTOR_TA_MGMT_PTA,CFG_WITH_USER_TA)
-
-ifeq ($(CFG_SECSTOR_TA_MGMT_PTA),y)
-$(call force,CFG_SECSTOR_TA,y,required by CFG_SECSTOR_TA_MGMT_PTA)
-endif
-
 
 # Setup compiler for this sub module
 COMPILER_$(sm)         ?= $(COMPILER)
diff --git a/mk/config.mk b/mk/config.mk
index 2c4e38377665..4835f98a2352 100644
--- a/mk/config.mk
+++ b/mk/config.mk
@@ -265,10 +265,12 @@ CFG_SECURE_DATA_PATH ?= n
 # Enable storage for TAs in secure storage, depends on CFG_REE_FS=y
 # TA binaries are stored encrypted in the REE FS and are protected by
 # metadata in secure storage.
-CFG_SECSTOR_TA ?= y
+CFG_SECSTOR_TA ?= $(call cfg-all-enabled,CFG_REE_FS CFG_WITH_USER_TA)
+$(eval $(call cfg-depends-all,CFG_SECSTOR_TA,CFG_REE_FS CFG_WITH_USER_TA))
 
 # Enable the pseudo TA that managages TA storage in secure storage
-CFG_SECSTOR_TA_MGMT_PTA ?= y
+CFG_SECSTOR_TA_MGMT_PTA ?= $(call cfg-all-enabled,CFG_SECSTOR_TA)
+$(eval $(call cfg-depends-all,CFG_SECSTOR_TA_MGMT_PTA,CFG_SECSTOR_TA))
 
 # Define the number of cores per cluster used in calculating core position.
 # The cluster number is shifted by this value and added to the core ID,
```

The first line (`cfg-all-enabled`) enables the feature by default if dependencies are met, while the second line (`cfg-depends-all`) disables it with a warning in case it is already set but dependencies are not met.
[We could as well fail the build instead of warn and disable ; and write a macro to encapsulate the whole pattern. But that is a topic for a separate PR].

Thoughts?

jenswi-linaro replied:
Thanks, I'll fix

jforissier commented on core/arch/arm/pta/secstor_ta_mgmt.c:
```diff
@@ -0,0 +1,202 @@
+				 uint32_t param_types,
+				 TEE_Param params[TEE_NUM_PARAMS])
+{
+	struct tee_ta_session *s = tee_ta_get_calling_session();
```
Unused => remove

jenswi-linaro commented:
Update

jenswi-linaro commented:
Are we OK with this? Or should I squash and rebase for one final round?

lorc commented:
I think it will be good to squash it, so we can review it patch-by-patch.

Also, looks like missed couple of my comments. This: https://github.com/OP-TEE/optee_os/pull/1928#discussion_r150818801
and this: https://github.com/OP-TEE/optee_os/pull/1928#discussion_r150821326

jforissier commented:
+1 for squash and rebase.

jenswi-linaro commented:
@lorc, sorry, addressed the comments now.

jenswi-linaro commented:
Squashed and rebased, for reference the old state was bef845dc00304248cdb2b4e38aa1d9c031c57232

jenswi-linaro commented:
Fixed checkpatch error

jenswi-linaro commented on core/crypto/signed_hdr.c:
```diff
@@ -0,0 +1,82 @@
+		return NULL;
+	memcpy(shdr, img, shdr_size);
+
+	/* Check that the data wasn't modified before the copy was completed */
```
If `shdr->hash_size` or `shdr->sig_size` is increased we may access data outside the allocated buffer later on.

jenswi-linaro commented:
`xtest _9` is failing because there's a secure storage object created for the TA database, this changes the names of the other secure storage objects. That's the main reason why I updated the htree tests a while ago, to be able to get rid of the `xtest _9` tests.

jenswi-linaro commented:
Comments, addressed, squashed and tags applied.

lorc commented:
`Reviewed-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>`

jenswi-linaro commented:
Tag applied

jenswi-linaro commented:
What's next step here?

jforissier commented:
Since this breaks the `xtest _9` series, I'd like to merge https://github.com/OP-TEE/optee_test/pull/240 first.

jforissier commented:
All good now. Thanks.
